File: FlexServ Backend/SpringBoot/AuthService/src/main/java/com/flexserv/controller/ServiceProviderController.java
package com.flexserv.controller;

import com.flexserv.dto.request.*;
import com.flexserv.dto.response.*;
import com.flexserv.service.ServiceProviderService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import com.flexserv.dto.request.UpdateServiceProviderRequestDto;

import java.util.List;
import java.util.Set;

@RestController
@RequestMapping("/api/providers")
@RequiredArgsConstructor
public class ServiceProviderController {

    private final ServiceProviderService providerService;

    @GetMapping
    public ResponseEntity<List<ServiceProviderResponseDto>> getAllProviders(
            @RequestParam(required = false) Long categoryId,
            @RequestParam(required = false) String keyword) {
        return ResponseEntity.ok(providerService.getAllProviders(categoryId, keyword));
    }


    @PostMapping
    public ResponseEntity<ServiceProviderResponseDto> createProvider(@RequestBody ServiceProviderRequestDto requestDto) {
        return new ResponseEntity<>(providerService.createProvider(requestDto), HttpStatus.CREATED);
    }

    @GetMapping("/{id}")
    public ResponseEntity<ServiceProviderResponseDto> getProviderById(@PathVariable Long id) {
        return ResponseEntity.ok(providerService.getProviderById(id));
    }

    @PutMapping("/{id}")
    public ResponseEntity<ServiceProviderResponseDto> updateProvider(
            @PathVariable Long id,
            @RequestBody UpdateServiceProviderRequestDto requestDto) {

        return ResponseEntity.ok(providerService.updateProvider(id, requestDto));
    }

    @GetMapping("/user/{userId}")
    public ResponseEntity<ServiceProviderResponseDto> getProviderByUserId(@PathVariable Long userId) {
        return ResponseEntity.ok(providerService.getProviderByUserId(userId));
    }

    @GetMapping("/available")
    public ResponseEntity<List<ServiceProviderResponseDto>> getAvailableProviders() {
        return ResponseEntity.ok(providerService.getAllAvailableProviders());
    }

    // @PatchMapping("/{id}/availability")
    // public ResponseEntity<ServiceProviderResponseDto> updateAvailability(
    //         @PathVariable Long id,
    //         @RequestParam Boolean available) {
    //     return ResponseEntity.ok(providerService.updateAvailability(id, available));
    // }

    @PostMapping("/{id}/services")
    public ResponseEntity<ServiceProviderResponseDto> addServices(
            @PathVariable Long id,
            @RequestBody Set<Long> serviceIds) {
        return ResponseEntity.ok(providerService.addServicesToProvider(id, serviceIds));
    }
}

File: FlexServ Backend/DotNet/AuthService/Controllers/ServiceProviderController.cs
using AuthService.Dtos;
using AuthService.Services;
using Microsoft.AspNetCore.Mvc;
using System;
using System.Collections.Generic;

namespace AuthService.Controllers
{
    [ApiController]
    [Route("api/providers")]
    public class ServiceProviderController : ControllerBase
    {
        private readonly IServiceProviderService _providerService;

        public ServiceProviderController(IServiceProviderService providerService)
        {
            _providerService = providerService;
        }

        [HttpGet]
        public IActionResult GetAllProviders([FromQuery] long? categoryId, [FromQuery] string? keyword)
        {
            var providers = _providerService.GetAllProviders(categoryId, keyword);
            return Ok(providers);
        }

        [HttpPost]
        public IActionResult CreateProvider([FromBody] ServiceProviderRequestDto requestDto)
        {
            var created = _providerService.CreateProvider(requestDto);
            return CreatedAtAction(nameof(GetProviderById), new { id = created.Id }, created);
        }

        [HttpGet("{id:long}")]
        public IActionResult GetProviderById(long id)
        {
            try
            {
                var provider = _providerService.GetProviderById(id);
                return Ok(provider);
            }
            catch (KeyNotFoundException ex)
            {
                return NotFound(ex.Message);
            }
        }

        [HttpPut("{id:long}")]
        public IActionResult UpdateProvider(long id, [FromBody] UpdateServiceProviderRequestDto requestDto)
        {
            try
            {
                var updated = _providerService.UpdateProvider(id, requestDto);
                return Ok(updated);
            }
            catch (KeyNotFoundException ex)
            {
                return NotFound(ex.Message);
            }
        }

        [HttpGet("user/{userId:long}")]
        public IActionResult GetProviderByUserId(long userId)
        {
            try
            {
                var provider = _providerService.GetProviderByUserId(userId);
                return Ok(provider);
            }
            catch (KeyNotFoundException ex)
            {
                return NotFound(ex.Message);
            }
        }

        [HttpGet("available")]
        public IActionResult GetAvailableProviders()
        {
            var providers = _providerService.GetAllAvailableProviders();
            return Ok(providers);
        }

        [HttpPost("{id:long}/services")]
        public IActionResult AddServices(long id, [FromBody] HashSet<long> serviceIds)
        {
            try
            {
                var updated = _providerService.AddServicesToProvider(id, serviceIds);
                return Ok(updated);
            }
            catch (KeyNotFoundException ex)
            {
                return NotFound(ex.Message);
            }
        }
    }
}
